# <a id='toc1_'></a>[Credit Card Fraud](#toc0_)

### <a id='toc1_1_1_'></a>[Introduction](#toc1_1_1_)      [&#8593;](#toc0_)

This dataset consists of credit card transactions in the western United States. It includes information about each transaction including customer details, the merchant and category of purchase, and whether or not the transaction was a fraud.

This analysis uses a combination of SQL and Tableau. Python will be used in the future to predict whether a transaction will be fraudulent.


The SQL statement retrieves a list of merchants from the transactions table, along with the count of fraudulent transactions associated with each merchant, and the RATIO of fraudulent to non-fraudulent transactions. The SQL statement outputs the 10 merchants with the largest fraudulent to non-fraudulent ratios - model could use ratios as a parameter to predict whether a transaction is fraudulent.

The second SQL statement is a simple query to find the merchants with the most amount of fraudulent transactions.


**Table of contents**<a id='toc0_'></a>    
- [Credit Card Fraud](#toc1_)    
    - [Introduction     ](#toc1_1_1_)    
    - [Convert CSV to .db file](#toc1_1_2_)    
    - [Load the SQL Extension and Database](#toc1_1_3_)    
    - [Data Table](#toc1_1_4_)    
      - [Data Dictionary](#toc1_1_4_1_)    
    - [Q1: Class Distribution](#toc1_1_5_)    
    - [Q2: Fraud Frequency by Merchant](#toc1_1_6_)    
    - [Q3: Top Categories Involved in Fraudulent Transactions](#toc1_1_7_)    
    - [Q4: Age of Cardholders Involved in Fraudulent Transactions](#toc1_1_8_)    
    - [Q5: Time periods where fraud is prevalent](#toc1_1_9_)    
    - [Q6: Average Transaction Amount for Fraud](#toc1_1_10_)    
    - [Q7: Specific cities where fraud is more prevalent?](#toc1_1_11_)    
    - [Q8: States that have the Highest Average Transaction Amounts in Fraudulent Cases](#toc1_1_12_)    

<!-- vscode-jupyter-toc-config
	numbering=false
	anchor=true
	flat=false
	minLevel=1
	maxLevel=6
	/vscode-jupyter-toc-config -->
<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->

### <a id='toc1_1_2_'></a>[Convert CSV to .db file](#toc0_)

In [ ]:
## Code block to convert csv file to .db using pandas

import pandas as pd
import sqlite3

# Load the CSV into a Pandas DataFrame
df = pd.read_csv('credit_card_fraud.csv')

# Create an SQLite database (or connect to an existing one)
conn = sqlite3.connect('credit_card_fraud.db')

# Write the data from the DataFrame to a new table in the SQLite database
df.to_sql('your_table_name', conn, if_exists='replace', index=False)

# Close the connection
conn.close()


### <a id='toc1_1_3_'></a>[Load the SQL Extension and Database](#toc0_)

In [32]:
from IPython.display import display, HTML
import pandas as pd

# Load the SQL extension
# %load_ext sql
%reload_ext sql

# Connect to your SQLite database
%sql sqlite:///C:/Users/Laxman/OneDrive/Documents/Portfolio/SQL/Projects/CreditCardFraud/credit_card_fraud.db

# Define a function to display a DataFrame in a scrollable format
def display_scrollable(df, height=400, width=800):
    """Display a DataFrame in a scrollable HTML div."""
    display(HTML(f"""
        <div style="overflow: auto; max-height: {height}px; max-width: {width}px; border: 1px solid black;">
            {df.to_html(index=False)}
        </div>
    """))


### <a id='toc1_1_4_'></a>[Data Table](#toc0_)

In [41]:
%%sql 
SELECT * FROM transactions
LIMIT 10;


 * sqlite:///C:/Users/Laxman/OneDrive/Documents/Portfolio/SQL/Projects/CreditCardFraud/credit_card_fraud.db
Done.


[('2019-01-01 00:00:44', 'Heller, Gutmann and Zieme', 'grocery_pos', 107.23, 'Orient', 'WA', 48.8878, -118.2105, 149, 'Special educational needs teacher', '1978-06-21', '1f76529f8574734946361c461b024d99', 49.159047, -118.186462, 0),
 ('2019-01-01 00:00:51', 'Lind-Buckridge', 'entertainment', 220.11, 'Malad City', 'ID', 42.1808, -112.262, 4154, 'Nature conservation officer', '1962-01-19', 'a1a22d70485983eac12b5b88dad1cf95', 43.150704, -112.154481, 0),
 ('2019-01-01 00:07:27', 'Kiehn Inc', 'grocery_pos', 96.29, 'Grenada', 'CA', 41.6125, -122.5258, 589, 'Systems analyst', '1945-12-21', '413636e759663f264aae1819a4d4f231', 41.65752, -122.230347, 0),
 ('2019-01-01 00:09:03', 'Beier-Hyatt', 'shopping_pos', 7.77, 'High Rolls Mountain Park', 'NM', 32.9396, -105.8189, 899, 'Naval architect', '1967-08-30', '8a6293af5ed278dea14448ded2685fea', 32.863258, -106.520205, 0),
 ('2019-01-01 00:21:32', 'Bruen-Yost', 'misc_pos', 6.85, 'Freedom', 'WY', 43.0172, -111.0292, 471, 'Education officer, museum', '1967-08-02', 'f3c43d336e92a44fc2fb67058d5949e3', 43.753735, -111.454923, 0),
 ('2019-01-01 00:22:06', 'Kunze Inc', 'grocery_pos', 90.22, 'Honokaa', 'HI', 20.0827, -155.488, 4878, 'Physiotherapist', '1966-12-03', '95826e3caa9e0b905294c6dae985aec1', 19.560013, -156.045889, 0),
 ('2019-01-01 00:22:18', 'Nitzsche, Kessler and Wol', 'shopping_pos', 4.02, 'Valentine', 'NE', 42.8062, -100.6215, 4005, 'Network engineer', '1945-03-15', '20490f3f0966ce74b4aaba8dc2c4ed52', 42.47559, -101.265846, 0),
 ('2019-01-01 00:22:36', 'Kihn, Abernathy and Douglas', 'shopping_net', 3.66, 'Westfir', 'OR', 43.7575, -122.481, 597, 'Forensic psychologist', '1961-05-19', '870c92b288a974a2faf1f24b05c27e33', 44.278191, -121.815161, 0),
 ('2019-01-01 00:31:51', 'Ledner-Pfannerstill', 'gas_transport', 102.13, 'Thompson', 'UT', 38.9999, -109.615, 46, 'Surveyor, minerals', '1987-04-23', '47238da5b40d126c8abea40a857c7809', 39.807313, -109.348294, 0),
 ('2019-01-01 00:34:10', 'Stracke-Lemke', 'grocery_pos', 83.07, 'Conway', 'WA', 48.34, -122.3456, 85, 'Research officer, political party', '1984-09-01', '9b7a0619dcc5c572dc134f2827ed5a6b', 48.682111, -122.719904, 0)]

#### <a id='toc1_1_4_1_'></a>[Data Dictionary](#toc0_)

| transdatetrans_time | Transaction DateTime                        |
|---------------------|---------------------------------------------|
| merchant            | Merchant Name                               |
| category            | Category of Merchant                        |
| amt                 | Amount of Transaction                       |
| city                | City of Credit Card Holder                  |
| state               | State of Credit Card Holder                 |
| lat                 | Latitude Location of Purchase               |
| long                | Longitude Location of Purchase              |
| city_pop            | Credit Card Holder's City Population        |
| job                 | Job of Credit Card Holder                   |
| dob                 | Date of Birth of Credit Card Holder         |
| trans_num           | Transaction Number                          |
| merch_lat           | Latitude Location of Merchant               |
| merch_long          | Longitude Location of Merchant              |
| is_fraud            | Whether Transaction is Fraud (1) or Not (0) |

### <a id='toc1_1_5_'></a>[Q1: Class Distribution](#toc0_)

In [16]:
%%sql
SELECT 
    COUNT(*) FILTER (WHERE is_fraud = 1) AS Fraud_Count,
    COUNT(*) FILTER (WHERE is_fraud = 0) AS Non_Fraud_Count
FROM transactions;


 * sqlite:///C:/Users/Laxman/OneDrive/Documents/Portfolio/SQL/Projects/CreditCardFraud/credit_card_fraud.db
Done.


[(1782, 337825)]

| not_fraud | is_fraud |
|-----------|----------|
| 337825    | 1782     |

This SQL query calculates the total number of non-fraudulent and fraudulent transactions in the transactions table. Ratio of non-fraudulent to fraudulent transactions is 169:1. Using SMOTE(Synthetic Minority Over-sampling Technique) in future studies might be beneficial when creating a model to predict fraudulent transactions to improve the model performance and as the imbalance can lead to biases. More research is required in SMOTE.

<img src="../../../Tableau/Project/CreditCardFraud/Q1 - Fraudulent vs  Non-Fraudulent.png" alt="" width="500" height="400">


### <a id='toc1_1_6_'></a>[Q2: Fraud Frequency by Merchant](#toc0_)

In [17]:
%%sql
SELECT 
    merchant AS Merchant,
    COUNT(*) FILTER (WHERE is_fraud = 1) AS Fraud_Count,
    COUNT(*) FILTER (WHERE is_fraud = 0) AS Non_Fraud_Count,
    ROUND(CAST(COUNT(*) FILTER (WHERE is_fraud = 1) AS FLOAT) / 
          CAST(COUNT(*) AS FLOAT), 4) AS Ratio_of_Transactions
FROM transactions
GROUP BY merchant
ORDER BY Ratio_of_Transactions DESC
LIMIT 10;



 * sqlite:///C:/Users/Laxman/OneDrive/Documents/Portfolio/SQL/Projects/CreditCardFraud/credit_card_fraud.db
Done.


[('Romaguera, Cruickshank and Greenholt', 18, 503, 0.0345),
 ('Kerluke-Abshire', 17, 498, 0.033),
 ('Kiehn-Emmerich', 19, 639, 0.0289),
 ('Terry-Huel', 13, 506, 0.025),
 ('Kunze Inc', 16, 629, 0.0248),
 ('Tillman, Fritsch and Schmitt', 9, 355, 0.0247),
 ('Moore, Dibbert and Koepp', 8, 329, 0.0237),
 ('Welch Inc', 8, 334, 0.0234),
 ('Lebsack and Sons', 8, 350, 0.0223),
 ('Kerluke Inc', 7, 319, 0.0215)]

The SQL statement retrieves a list of merchants from the transactions table, along with the count of fraudulent transactions associated with each merchant, and the RATIO of fraudulent to non-fraudulent transactions. The SQL statement outputs the 10 merchants with the largest fraudulent to non-fraudulent ratios - model could use ratios as a parameter to predict whether a transaction is fraudulent.

The second SQL statement is a simple query to find the merchants with the most amount of fraudulent transactions.


In [18]:
%%sql
SELECT 
    merchant AS Merchant,
    COUNT(*) FILTER (WHERE is_fraud = 1) AS Fraud_Count,
    COUNT(*) FILTER (WHERE is_fraud = 0) AS Non_Fraud_Count,
    ROUND(CAST(COUNT(*) FILTER (WHERE is_fraud = 1) AS FLOAT) / 
          CAST(COUNT(*) AS FLOAT), 4) AS Ratio_of_Transactions
FROM transactions
GROUP BY merchant
ORDER BY Fraud_Count DESC
LIMIT 10;


 * sqlite:///C:/Users/Laxman/OneDrive/Documents/Portfolio/SQL/Projects/CreditCardFraud/credit_card_fraud.db
Done.


[('Kiehn-Emmerich', 19, 639, 0.0289),
 ('Romaguera, Cruickshank and Greenholt', 18, 503, 0.0345),
 ('Kerluke-Abshire', 17, 498, 0.033),
 ('Kunze Inc', 16, 629, 0.0248),
 ('Kilback LLC', 15, 1134, 0.0131),
 ('Strosin-Cruickshank', 14, 657, 0.0209),
 ('Terry-Huel', 13, 506, 0.025),
 ('Schultz, Simonis and Little', 13, 609, 0.0209),
 ('Murray-Smitham', 13, 651, 0.0196),
 ('McDermott-Weimann', 13, 626, 0.0203)]

The SQL statement retrieves a list of merchants from the transactions table, along with the count of fraudulent transactions associated with each merchant, and the RATIO of fraudulent to non-fraudulent transactions. The SQL statement outputs the 10 merchants with the largest fraudulent to non-fraudulent ratios - model could use ratios as a parameter to predict whether a transaction is fraudulent.

The second SQL statement is a simple query to find the merchants with the most amount of fraudulent transactions.


### <a id='toc1_1_7_'></a>[Q3: Top Categories Involved in Fraudulent Transactions](#toc0_)

In [19]:
%%sql
SELECT 
    category AS Category,
    COUNT(*) FILTER (WHERE is_fraud = 1) AS Fraud_Count,
    COUNT(*) FILTER (WHERE is_fraud = 0) AS Non_Fraud_Count,
    ROUND(CAST(COUNT(*) FILTER (WHERE is_fraud = 1) AS FLOAT) / 
          CAST(COUNT(*) AS FLOAT), 4) AS Ratio_of_Transactions
FROM transactions
GROUP BY Category
ORDER BY Ratio_of_Transactions DESC
LIMIT 10;


 * sqlite:///C:/Users/Laxman/OneDrive/Documents/Portfolio/SQL/Projects/CreditCardFraud/credit_card_fraud.db
Done.


[('shopping_net', 381, 25998, 0.0144),
 ('grocery_pos', 433, 32299, 0.0132),
 ('misc_net', 217, 16681, 0.0128),
 ('shopping_pos', 187, 30142, 0.0062),
 ('gas_transport', 153, 34936, 0.0044),
 ('travel', 33, 10289, 0.0032),
 ('misc_pos', 62, 19962, 0.0031),
 ('grocery_net', 27, 11328, 0.0024),
 ('personal_care', 55, 24351, 0.0023),
 ('entertainment', 55, 24167, 0.0023)]

Both important: 
    - Ratios: making it easier to focus resources on those high-risk categories.
    - Count: High Counts indicates categories require more attention due to sheer volume of fraud.
  
From the query: 
- shopping_net is the most high-risk category with the largest ratio with next largest grocery_pos
- grocery_pos has the most fraud with shopping_net second

### <a id='toc1_1_8_'></a>[Q4: Age of Cardholders Involved in Fraudulent Transactions](#toc0_)

In [20]:
%%sql
SELECT 
    FLOOR((strftime('%Y', 'now') - strftime('%Y', dob)) / 10) * 10 AS Age_Bin_Start,
    FLOOR((strftime('%Y', 'now') - strftime('%Y', dob)) / 10) * 10 + 9 AS Age_Bin_End,
    COUNT(*) FILTER (WHERE is_fraud = 1) AS Fraud_Count
FROM transactions
GROUP BY Age_Bin_Start, Age_Bin_End
ORDER BY Fraud_Count DESC
LIMIT 10;


 * sqlite:///C:/Users/Laxman/OneDrive/Documents/Portfolio/SQL/Projects/CreditCardFraud/credit_card_fraud.db
Done.


[(50, 59, 432),
 (60, 69, 315),
 (30, 39, 283),
 (40, 49, 241),
 (80, 89, 168),
 (20, 29, 149),
 (70, 79, 135),
 (90, 99, 59)]

Target Age Groups: 50 to 70 are most targeted.
could be explained with out of touch from technology while more > 70 use cash to pay. Need a cumulative graph to say with certain

Information could be used to focus prevention efforts at certain age groups.


In [21]:
%%sql
SELECT 
    CAST(FLOOR((strftime('%Y', 'now') - strftime('%Y', dob)) / 10) * 10 AS TEXT) || ' - ' || 
    CAST((FLOOR((strftime('%Y', 'now') - strftime('%Y', dob)) / 10) * 10 + 9) AS TEXT) AS Age_Bin,
    COUNT(*) FILTER (WHERE is_fraud = 1) AS Fraud_Count,
    COUNT(*) FILTER (WHERE is_fraud = 0) AS Non_Fraud_Count,
    ROUND(CAST(COUNT(*) FILTER (WHERE is_fraud = 1) AS FLOAT) / 
          CAST(COUNT(*) AS FLOAT), 4) AS Ratio_of_Transactions_Fraudulent
FROM transactions
GROUP BY Age_Bin
ORDER BY Ratio_of_Transactions_Fraudulent DESC
LIMIT 5;


 * sqlite:///C:/Users/Laxman/OneDrive/Documents/Portfolio/SQL/Projects/CreditCardFraud/credit_card_fraud.db
Done.


[('20 - 29', 149, 17473, 0.0085),
 ('80 - 89', 168, 20384, 0.0082),
 ('60 - 69', 315, 40051, 0.0078),
 ('90 - 99', 59, 9459, 0.0062),
 ('70 - 79', 135, 24755, 0.0054)]

Include Total Transaction in the analysis as it includes the overall transaction activity of the age group. 

Research suggests that people under 35
are more likely than older age groups to have
been targeted in an impersonation scam and
be swayed to provide personal or financial
information. - [Reference](https://www.ukfinance.org.uk/system/files/2023-05/Annual%20Fraud%20Report%202023_0.pdf)

### <a id='toc1_1_9_'></a>[Q5: Time periods where fraud is prevalent](#toc0_)

In [22]:
%%sql
SELECT 
    CAST(FLOOR(strftime('%H', trans_date_trans_time) / 6) * 6 AS TEXT) || ' - ' || 
    CAST(FLOOR(strftime('%H', trans_date_trans_time) / 6) * 6 + 6 AS TEXT) As Hour_Bin,
    COUNT(*) FILTER (WHERE is_fraud = 1) AS Fraud_Count,
    COUNT(*) FILTER (WHERE is_fraud = 0) AS Non_Fraud_Count,
    ROUND(CAST(COUNT(*) FILTER (WHERE is_fraud = 1) AS FLOAT) / 
          CAST(COUNT(*) AS FLOAT), 4) AS Ratio_of_Transactions_Fraudulent
FROM transactions
Group By Hour_Bin
ORDER BY Ratio_of_Transactions_Fraudulent DESC
LIMIT 5;


 * sqlite:///C:/Users/Laxman/OneDrive/Documents/Portfolio/SQL/Projects/CreditCardFraud/credit_card_fraud.db
Done.


[('0 - 6', 636, 65817, 0.0096),
 ('18 - 24', 980, 102859, 0.0094),
 ('12 - 18', 111, 102706, 0.0011),
 ('6 - 12', 55, 66443, 0.0008)]

Analysis:
- The higher fraud ratios during the late evening (18 - 24) and early morning (0 - 6) might indicate that fraudsters are more active during these times, likely due to lower vigilance from both individuals and financial institutions. 
- 6 to 18 are peak hours and might have more monitoring by banks.
- 18 to 6 -> fraudsters are more likely to attempt fraudulent transactions when individuals are either asleep


### <a id='toc1_1_10_'></a>[Q6: Average Transaction Amount for Fraud](#toc0_)

In [23]:
%%sql
SELECT 
    CAST(FLOOR(amt / 100) * 100 AS TEXT) || ' - ' || 
    CAST(FLOOR(amt / 100) * 100 + 100 AS TEXT) As Transaction_Amount_Bin,
    COUNT(*) FILTER (WHERE is_fraud = 1) AS Fraud_Count,
    COUNT(*) FILTER (WHERE is_fraud = 0) AS Non_Fraud_Count,
    ROUND(CAST(COUNT(*) FILTER (WHERE is_fraud = 1) AS FLOAT) / 
          CAST(COUNT(*) AS FLOAT), 4) AS Ratio_of_Transactions_Fraudulent
FROM transactions
Group By Transaction_Amount_Bin
Having Fraud_Count > 0
ORDER BY Ratio_of_Transactions_Fraudulent DESC;


 * sqlite:///C:/Users/Laxman/OneDrive/Documents/Portfolio/SQL/Projects/CreditCardFraud/credit_card_fraud.db
Done.


[('900 - 1000', 209, 175, 0.5443),
 ('1000 - 1100', 146, 136, 0.5177),
 ('800 - 900', 192, 234, 0.4507),
 ('1100 - 1200', 61, 108, 0.3609),
 ('700 - 800', 145, 352, 0.2918),
 ('300 - 400', 303, 1988, 0.1323),
 ('1200 - 1300', 9, 93, 0.0882),
 ('600 - 700', 48, 510, 0.086),
 ('1300 - 1400', 3, 64, 0.0448),
 ('500 - 600', 24, 1081, 0.0217),
 ('200 - 300', 180, 8631, 0.0204),
 ('400 - 500', 18, 1532, 0.0116),
 ('0 - 100', 407, 276313, 0.0015),
 ('100 - 200', 37, 46154, 0.0008)]

Analysis:
- High fraud ratios in large transaction amounts
- Clear trend from large to small transaction amounts - offer smaller financial rewards to fraudsters.
- 0-300 has a high volume of Fraud - fraudsters are still active in small transactions,
- Fraudsters are targeting 700-1100 fraud ratios peaking at over 50% in the $900 - $1100 bins, and 0-100 targeted due to the high-volume of transactions.

### <a id='toc1_1_11_'></a>[Q7: Specific cities where fraud is more prevalent?](#toc0_)


In [24]:
%%sql
SELECT 
    city AS City,
    COUNT(*) FILTER (WHERE is_fraud = 1) AS Fraud_Count,
    COUNT(*) FILTER (WHERE is_fraud = 0) AS Non_Fraud_Count,
    ROUND(CAST(COUNT(*) FILTER (WHERE is_fraud = 1) AS FLOAT) / 
          CAST(COUNT(*) AS FLOAT), 4) AS Ratio_of_Transactions_Fraudulent,
    city_pop      
FROM transactions
Group By City
ORDER BY Ratio_of_Transactions_Fraudulent DESC
LIMIT 5;


 * sqlite:///C:/Users/Laxman/OneDrive/Documents/Portfolio/SQL/Projects/CreditCardFraud/credit_card_fraud.db
Done.


[('Wappapello', 8, 0, 1.0, 2601),
 ('Vacaville', 7, 0, 1.0, 99475),
 ('Seligman', 10, 0, 1.0, 2693),
 ('Seattle', 19, 0, 1.0, 837792),
 ('Pleasant Hill', 8, 0, 1.0, 12866)]

Analysis:
- Certain cities only have fraud - possible fraud targeting snaller cities
- Data might not be complete for example Seattle has 0 Non fraud Counts while having a population of a 1million



In [25]:
%%sql
SELECT 
    city AS City,
    COUNT(*) FILTER (WHERE is_fraud = 1) AS Fraud_Count,
    COUNT(*) FILTER (WHERE is_fraud = 0) AS Non_Fraud_Count,
    city_pop      
FROM transactions
Group By City
ORDER BY Fraud_Count DESC
LIMIT 5;


 * sqlite:///C:/Users/Laxman/OneDrive/Documents/Portfolio/SQL/Projects/CreditCardFraud/credit_card_fraud.db
Done.


[('Albuquerque', 24, 1455, 641349),
 ('Aurora', 23, 727, 389246),
 ('Fort Washakie', 21, 5095, 1645),
 ('Mesa', 20, 2910, 129),
 ('Glendale', 20, 3639, 172817)]

### <a id='toc1_1_12_'></a>[Q8: States that have the Highest Average Transaction Amounts in Fraudulent Cases](#toc0_)

In [26]:
%%sql
SELECT 
    state AS State,
    COUNT(*) FILTER (WHERE is_fraud = 1) AS Fraud_Count,
    COUNT(*) FILTER (WHERE is_fraud = 0) AS Non_Fraud_Count,
    ROUND(CAST(COUNT(*) FILTER (WHERE is_fraud = 1) AS FLOAT) / 
          CAST(COUNT(*) AS FLOAT), 4) AS Ratio_of_Transactions_Fraudulent
FROM transactions
Group By State
ORDER BY Ratio_of_Transactions_Fraudulent DESC
LIMIT 15;


 * sqlite:///C:/Users/Laxman/OneDrive/Documents/Portfolio/SQL/Projects/CreditCardFraud/credit_card_fraud.db
Done.


[('AK', 50, 2913, 0.0169),
 ('OR', 197, 26211, 0.0075),
 ('NE', 216, 34209, 0.0063),
 ('CO', 115, 19651, 0.0058),
 ('NM', 121, 23306, 0.0052),
 ('CA', 402, 80093, 0.005),
 ('MO', 262, 54642, 0.0048),
 ('WA', 126, 26914, 0.0047),
 ('HI', 16, 3633, 0.0044),
 ('WY', 119, 27657, 0.0043),
 ('AZ', 64, 15298, 0.0042),
 ('ID', 33, 8002, 0.0041),
 ('UT', 61, 15296, 0.004)]

Analysis:
- AK has a high number of fraud
